src/passes/TranslateEH.cpp

```diff
@@ -0,0 +1,822 @@
+
+  // Scans and records which try labels are targeted by delegates and rethrows.
+  struct TargetTryLabelScanner : public PostWalker<TargetTryLabelScanner> {
+    TargetTryLabelScanner(Function* func) { walkFunction(func); }
+
+    std::unordered_set<Name> delegateTargets;
+    std::unordered_set<Name> rethrowTargets;
+    bool isTargetedByDelegates(Try* curr) const {
+      return delegateTargets.find(curr->name) != delegateTargets.end();
+    }
+    bool isTargetedByRethrows(Try* curr) const {
+      return rethrowTargets.find(curr->name) != rethrowTargets.end();
+    }
+
+    void visitTry(Try* curr) {
+      if (curr->isDelegate()) {
+        delegateTargets.insert(curr->delegateTarget);
+      }
+    }
+
+    void visitRethrow(Rethrow* curr) { rethrowTargets.insert(curr->target); }
+  };
+
+  // For each try label targeted by rethrows, assign a local that will contain
+  // its exnref in the new spec (via try + catch(_all)_ref instruction), so that
+  // every 'rethrow $somelabel' can later be translated into
+  // 'throw_ref $somelocal'.
+  //
+  // In the examples below, try's (do) bodies are omitted for the sake of
+  // simplicity.
+  //
+  // For example,
+  // (try $l0
+  //   (catch
+  //     (try $l1
+  //       (catch
+  //         (rethrow $l0) ;; will become (throw_ref $local0)
+  //       )
+  //       (catch
+  //         (rethrow $l1) ;; will become (throw_ref $local1)
+  //       )
+  //     )
+  //   )
+  // )
+  // In this case, we need two locals for exnrefs, each for $l0 and $l1.
+  //
+  // Note that the number of locals required is the maximum depth of try-catch
+  // nests, when only counting 'try's that are targeted by rethrows, which means
+  // the pattern below only needs one extra local:
+  // (try $l0
+  //   (catch
+  //     (rethrow $l0)
+  //   )
+  // )
+  // (try $l1
+  //   (catch
+  //     (rethrow $l1)
+  //   )
+  // )
+  // Because the two trys are not nested within each other, one local can be
+  // reused in both trys.
+  //
+  // Also even if there is a try-catch nest with depth N, if only some of them
+  // are targeted by rethrows, we only need that many extra locals for the whole
+  // nest:
+  // (try $l0
+  //   (catch
+  //     (try $l1
+  //       (catch
+  //         ...
+  //         ...
+  //         (try $lN
+  //           (catch
+  //             ;; If this is the only rethrow in this nest, we only need one
+  //             ;; extra exnref local for all N try-catches
+  //             (rethrow $l1)
+  //           )
+  //         )
+  //         ...
+  //         ...
+  //       )
+  //     )
+  //   )
+  // )
+  struct ExnrefLocalAssigner : public PostWalker<ExnrefLocalAssigner> {
+    TargetTryLabelScanner* labelScanner = nullptr;
+
+    std::vector<Index> exnrefLocals;
+    std::unordered_map<Name, Index> rethrowTargetToExnrefLocal;
+
+    // Depth of the current try nest, when only counting trys targeted with
+    // rethrows.
+    size_t rethrowTryDepth = 0;
```

Comment:
True, but I'm not sure we can run `coalesce-locals` or `reorder-locals` with the new EH at this point yet... Supporting it in them wouldn't be a ton of work though. I wanted to make the translator emit maybe not the best but ok code at this point without spending more time on the optimization pipeline.

Also if we assign a local to each `try` the number of locals can be prohibitive. `CoalesceLocal.cpp` says https://github.com/WebAssembly/binaryen/blob/2dfa01f09071ab78e8134fee5c96bcd022330635/src/passes/CoalesceLocals.cpp#L22-L25, which means we should run `simplify-locals` too? 

Anyway, given that the optimization pipeline is not ready for the new EH yet, and also given that this local assignment in this pass is not very complicated, I am inclined to keep this at least for the moment, even though maybe we can delete that later after the opt pipeline is ready. WDYT?

Reply:
sgtm, this seems fine for now.

About that comment, in general we'd need `simplify-locals`, but I think the locals added in this pass will not benefit from that particular pass just because of the form they are in. I could be wrong though...